datumaro/components/format_detection.py

```diff
@@ -84,6 +85,27 @@ def root_path(self) -> str:
         """
         return self._root_path
 
+    def _is_path_within_root(self, path: str) -> bool:
+        """
+        Checks that `path` is a relative path and does not attempt to leave
+        the dataset root by using `..` segments.
+
+        Requirement-placing methods that use this to verify their arguments
+        should raise a FormatRequirementsUnmet rather than a "hard" error like
+        AssertionError if False is returned. The reason is that the path passed
+        by the detector might not have been hardcoded, and instead might have
+        been acquired from another file in the dataset. In that case, an invalid
+        pattern signifies a problem with the dataset, not with the detector.
```

Comment:
Maybe we need to raise invalid requirement error in such cases?

Reply:
What do you mean by that? A new exception type? How would it be handled?

Reply:
Raising `FormatRequirementsUnmet` doesn't seem correct when incorrect requirements are specified. But I agree that the whole detection process shouldn't be interrupted in such cases. So, probably, just another error can be introduced. I suppose, we can come up with:
`DatasetDetectionError` 
^- `FormatRequirementsUnmet` 
^- `InvalidRequirement`

And just catch and return `DatasetDetectionError`s.

Reply:
> So, probably, just another error can be introduced.

That will cause a problem later with the addition of alternatives. If one alternative raises a `FormatRequirementsUnmet`, and another raises an `InvalidRequirement`, then what exception should the detector as a whole raise?

Furthermore, I predict that "incorrect requirements" will, in practice, be caused by the dataset not meeting other format requirements (like in the scenario explained in the comment: when a value is read from a file that's supposed to be a path, but isn't), and therefore it's actually reasonable to report them as unmet requirements.

The only other possible cause that I can see would be an invalid requirement that is hardcoded into the detector, but such a detector will always fail, so an error like that would be caught in testing and corrected.

Reply:
> That will cause a problem later with the addition of alternatives. If one alternative raises a `FormatRequirementsUnmet`, and another raises an `InvalidRequirement`, then what exception should the detector as a whole raise?

I don't think I understand the way you want to implement alternatives, but unless they are implemented, the question makes no sense. From my perspective, we will iterate over alternatives and then return a list of errors. If there is a matching one, the confidence is returned. 

Maybe, if there are incorrect requirements, we need to stop checking the format at all. Print a debug log message / a warning, return the NONE confidence. I expect such situations to happen only in custom plugins, or during development. So maybe, we can just fail, actually.